lib/vrl/stdlib/src/parse_cef.rs

```diff
@@ -0,0 +1,494 @@
+                ),
+            },
+            Example {
+                title: "with syslog prefix",
+                source: r#"parse_cef!("Sep 29 08:26:10 host CEF:1|Security|threatmanager|1.0|100|worm successfully stopped|10|")"#,
+                result: Ok(
+                    r#"{"cefVersion":"1","deviceVendor":"Security","deviceProduct":"threatmanager","deviceVersion":"1.0","deviceEventClassId":"100","name":"worm successfully stopped","severity":"10"}"#,
+                ),
+            },
+            Example {
+                title: "escapes",
+                source: r#"parse_cef!(s'CEF:0|security|threatmanager|1.0|100|Detected a \| in message. No action needed.|10|src=10.0.0.1 msg=Detected a threat.\n No action needed act=blocked a \= dst=1.1.1.1')"#,
+                result: Ok(
+                    r#"{"cefVersion":"0","deviceVendor":"security","deviceProduct":"threatmanager","deviceVersion":"1.0","deviceEventClassId":"100","name":"Detected a | in message. No action needed.","severity":"10","src":"10.0.0.1","msg":"Detected a threat.\n No action needed","act":"blocked a =", "dst":"1.1.1.1"}"#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(ParseCefFn { value }.as_expr())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct ParseCefFn {
+    pub(crate) value: Box<dyn Expression>,
+}
+
+impl FunctionExpression for ParseCefFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let bytes = self.value.resolve(ctx)?;
+        let bytes = bytes.try_bytes_utf8_lossy()?;
+
+        parse(&bytes).map(Iterator::collect)
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {
+        type_def()
+    }
+}
+
+fn parse(input: &str) -> Result<impl Iterator<Item = (String, Value)> + '_> {
+    let (rest, (header, mut extension)) =
+        pair(parse_header, parse_extension)(input).map_err(|e| match e {
+            nom::Err::Error(e) | nom::Err::Failure(e) => {
+                // Create a descriptive error message if possible.
+                nom::error::convert_error(input, e)
+            }
+            nom::Err::Incomplete(_) => e.to_string(),
+        })?;
+
+    // Trim trailing whitespace on last extension value
+    if let Some((_, value)) = extension.last_mut() {
+        let suffix = value.trim_end_matches(' ');
+        value.truncate(suffix.len());
+    }
+
+    if rest.trim().is_empty() {
+        let result = [
+            "cefVersion",
+            "deviceVendor",
+            "deviceProduct",
+            "deviceVersion",
+            "deviceEventClassId",
+            "name",
+            "severity",
+        ]
+        .into_iter()
+        .zip(header)
+        .chain(extension)
```

Comment:
A bit pedantic, but one edge case I noticed is if an extension has the name of one of the "header" fields, the extension value is used instead. I'm not sure if this is important enough to make this result in an _error_, but a simple improvement would be to reverse the order of the iterator so the extension comes first, which will guarantee the "header" values are used over potential colliding extension values.